crypto/crypto.go

```diff
@@ -1,128 +1,68 @@
-// toECDSA creates a private key with the given D value. The strict parameter
-// controls whether the key's length should be enforced at the curve size or
-// it can also accept legacy encodings (0 prefixes).
-func toECDSA(d []byte, strict bool) (*ecdsa.PrivateKey, error) {
-	priv := new(ecdsa.PrivateKey)
-	priv.PublicKey.Curve = S256()
-	if strict && 8*len(d) != priv.Params().BitSize {
-		return nil, fmt.Errorf("invalid length, need %d bits", priv.Params().BitSize)
-	}
-	priv.D = new(big.Int).SetBytes(d)
-
-	// The priv.D must < N
-	if priv.D.Cmp(secp256k1N) >= 0 {
-		return nil, fmt.Errorf("invalid private key, >=N")
-	}
-	// The priv.D must not be zero or negative.
-	if priv.D.Sign() <= 0 {
-		return nil, fmt.Errorf("invalid private key, zero or negative")
-	}
-
-	priv.PublicKey.X, priv.PublicKey.Y = priv.PublicKey.Curve.ScalarBaseMult(d)
-	if priv.PublicKey.X == nil {
-		return nil, errors.New("invalid private key")
-	}
-	return priv, nil
+// Sign signs the given message, which must be 32 bytes long.
+func Sign(sk, msg []byte) ([]byte, error) {
+	return secp256k1.Sign(msg, sk)
 }
 
-// ECDSAToBytes exports a private key into a binary dump.
-func ECDSAToBytes(priv *ecdsa.PrivateKey) []byte {
-	if priv == nil {
-		return nil
-	}
-	return cu.PaddedBigBytes(priv.D, priv.Params().BitSize/8)
+// Equals compares two private key for equality and returns true if they are the same.
+func Equals(sk, other []byte) bool {
+	return bytes.Equal(sk, other)
 }
 
-// BytesToECDSAPub creates a public key with the given `pub` value.
-func BytesToECDSAPub(pub []byte) *ecdsa.PublicKey {
-	if len(pub) == 0 {
-		return nil
+// Verify checks the given signature and returns true if it is valid.
+func Verify(pk, msg, signature []byte) bool {
+	if len(signature) == 65 {
+		// Drop the V in [R | S | V] style signatures
+		return secp256k1.VerifySignature(pk[:], msg, signature[:len(signature)-2])
 	}
-	x, y := elliptic.Unmarshal(S256(), pub)
-	return &ecdsa.PublicKey{Curve: S256(), X: x, Y: y}
-}
 
-// ECDSAPubToBytes marshals `pub` to a slice of bytes
-func ECDSAPubToBytes(pub *ecdsa.PublicKey) []byte {
-	if pub == nil || pub.X == nil || pub.Y == nil {
-		return nil
-	}
-	return elliptic.Marshal(S256(), pub.X, pub.Y)
+	return secp256k1.VerifySignature(pk[:], msg, signature)
 }
 
-// HexToECDSA parses a secp256k1 private key.
-func HexToECDSA(hexkey string) (*ecdsa.PrivateKey, error) {
-	b, err := hex.DecodeString(hexkey)
+// GenerateKeyFromSeed generates a new key from the given reader.
+func GenerateKeyFromSeed(seed io.Reader) ([]byte, error) {
+	key, err := ecdsa.GenerateKey(secp256k1.S256(), seed)
 	if err != nil {
-		return nil, errors.New("invalid hex string")
+		return nil, err
 	}
-	return BytesToECDSA(b)
-}
 
-// GenerateKey generates an ecdsa private key
-func GenerateKey() (*ecdsa.PrivateKey, error) {
-	return GenerateKeyFromSeed(rand.Reader)
+	privkey := make([]byte, PrivateKeyBytes)
+	blob := key.D.Bytes()
+
+	copy(privkey[PrivateKeyBytes-len(blob):], blob)
```

Comment:
Please document the context/intent for this. I don't understand why it's needed. When could it blob not be the right length? What if blob is longer than PrivateKeyBytes?

Reply:
@dignifiedquire can you chime in on this one? I am unsure.

Reply:
Unless the elliptic curve implementation is broken this has a fixed size, always as the serialization for these points is defined to be a fixed byte size for this particular curve

Reply:
Ok. It would help less familiar reads to leave a comment here to that effect.